Modules/vc-module-catalog/VirtoCommerce.CatalogModule.Data/ExportImport2/CatalogFullExportPagedDataSource.cs

```diff
@@ -0,0 +1,211 @@
+
+        private bool _statesInitialized = false;
+
+        private readonly ExportDataSourceState[] _exportDataSourceStates = new ExportDataSourceState[5];
+
+        private ExportDataQuery _dataQuery;
+
+        public ExportDataQuery DataQuery
+        {
+            set
+            {
+                _dataQuery = value;
+                CurrentPageNumber = 0;
+                TotalCount = -1;
+                InitExportDataSourceStates();
+                CalculateCounts();
+            }
+        }
+
+        public int GetTotalCount()
+        {
+            CalculateCounts();
+            return TotalCount;
+        }
+
+        private void EnsureHaveTotals()
+        {
+            if (TotalCount < 0)
+            {
+                CalculateCounts();
+            }
+        }
+
+        public bool Fetch()
+        {
+            EnsureHaveTotals();
+
+            var take = _dataQuery.Take ?? PageSize;
+            var skip = _dataQuery.Skip ?? CurrentPageNumber * PageSize;
+            var taskList = new List<Task>();
+            Items = new List<IExportable>();
+
+            var curStateStartPos = 0; // Starting position of current state
+
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.FetchResult.Result = Array.Empty<IExportable>();
+                var overlap = skip - curStateStartPos; // Positive overlap should fall into skip, negative into take
+                state.SearchCriteria.Skip = overlap < 0 ? 0 : overlap;
+                state.SearchCriteria.Take = (skip + take > curStateStartPos + state.FetchResult.TotalCount) ? curStateStartPos + state.FetchResult.TotalCount - skip : take + (overlap < 0 ? overlap : 0);
+
+                if (state.SearchCriteria.Take > 0) // Skip running this fetch in case of nothing to take
+                {
+                    taskList.Add(state.FetchFunc(state));
+                }
+
+                curStateStartPos += state.FetchResult.TotalCount;
+            }
+
+            Task.WhenAll(taskList).GetAwaiter().GetResult(); // Ensures completing all fetches
+
+            ((List<IExportable>)Items).AddRange(_exportDataSourceStates.SelectMany(x => x.FetchResult.Result));
+            CurrentPageNumber++;
+
+            return Items.Any();
+        }
+
+        private void CalculateCounts()
+        {
+            var taskList = new List<Task>();
+            foreach (var state in _exportDataSourceStates)
+            {
+                state.SearchCriteria.Skip = 0;
+                state.SearchCriteria.Take = 0;
+                taskList.Add(state.FetchFunc(state));
+            }
+
+            Task.WhenAll(taskList).GetAwaiter().GetResult(); // Ensures completing all fetches
+            TotalCount = _exportDataSourceStates.Sum(x => x.FetchResult.TotalCount);
+        }
+
+        private void InitExportDataSourceStates()
+        {
+            if (!_statesInitialized)
```

Comment:
Why do we need this one? It is non shared class with this method called one inside constructor.

Reply:
Removed. Incapsulated into DataQuery setter.